src/sps.rs

```diff
@@ -0,0 +1,45 @@
+use crate::constants::NUM_CHALLENGE_BITS;
+use crate::plonk::{eval::Error as EvalError, PlonkInstance};
+use crate::poseidon::ROTrait;
+use crate::util::fe_to_fe;
+use halo2_proofs::arithmetic::CurveAffine;
+
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum Error {
+    #[error(transparent)]
+    Eval(#[from] EvalError),
+    #[error("Sps verification fail challenge not match at index {challenge_index}")]
+    ChallengeNotMatch { challenge_index: usize },
+    #[error("For this challenges count table must have lookup aguments")]
+    LackOfLookupArguments,
+    #[error("Lack of advices, should call `TableData::assembly` first")]
+    LackOfAdvices,
+    #[error("Only 0..=3 num of challenges supported: {challenges_count} not")]
+    UnsupportedChallengesCount { challenges_count: usize },
+}
+
+/// This trait verifies whether the instance is faithly generated by a Special soundness protocol (sps)
```

Comment:
referenses to protocol math-description

Reply:
added paper reference because the definition of the protocol is lengthy.

Reply:
But not general referense, please specify place in pdf

It's not a formal reference so that would be. It is a shortcut that will allow any developer to quickly understand where to look at the specification and understand where there may be an error, for example.